easydock/database.py

```diff
@@ -417,6 +417,156 @@ def add_protonation(db_fname, program='chemaxon', tautomerize=True, table_name='
     '''
     conn = sqlite3.connect(db_fname)
 
+    try:
+        cur = conn.cursor()
+
+        # SMiLES only
+        sql = f"""SELECT smi, id || '_' || stereo_id 
+                  FROM {table_name} 
+                  WHERE smi IS NOT NULL AND docking_score is NULL AND smi_protonated is NULL AND source_mol_block is NULL """
+        sql += add_sql
+        data_list_smi = list(cur.execute(sql))
+
+        # mol_block only
+        sql = f"""SELECT smi, id || '_' || stereo_id 
+                  FROM {table_name} 
+                  WHERE smi IS NOT NULL AND docking_score is NULL AND smi_protonated is NULL AND source_mol_block is NOT NULL """
+        sql += add_sql
+        data_list_mol = list(cur.execute(sql))
+
+        if not data_list_smi and not data_list_mol:
+            sys.stderr.write(f'no molecules to protonate\n')
+            return
+
+        smi_names = set(mol_name for smi, mol_name in data_list_smi)
+        mol_names = set(mol_name for smi, mol_name in data_list_mol)
+
+        output_data_smi = []
+        output_data_mol = []
+
+        if program in ['chemaxon']:  # first protocol
+            nmols = ncpu * 10  # to have batches of a reasonable size; the N value can be increased if necessary
+            for chunk in split_generator_to_chunks((data for data in data_list_mol + data_list_smi), nmols):
+                if chunk:
+                    with tempfile.NamedTemporaryFile(suffix='.smi', mode='w', encoding='utf-8') as tmp:
+                        for smi, mol_name in chunk:
+                            tmp.write(f'{smi}\t{mol_name}\n')
+                        tmp.flush()
+
+                    fd, output = tempfile.mkstemp()
+                    try:
+                        protonate_func = partial(protonate_chemaxon, tautomerize=tautomerize)
+                        read_func = read_protonate_chemaxon
+                        protonate_func(tmp.name, output)
+                        mols = read_func(output)
+                    finally:
+                        os.remove(output)
+                        os.close(fd)
```

Comment:
Move this block under with statement, otherwise tmp var will not be available for protonation function